src/utils/attempt.js

```diff
@@ -0,0 +1,11 @@
+// Given a `fallback` and a `preferred` function,
+// return a function that attempts to call the `preferred`.
+// If it fails, or returns undefined, call the fallback and return its value.
+export default function attempt ( fallback, preferred ) {
+	return function ( ...args ) {
+		const boundFallback = () => fallback( ...args );
+
+		return Promise.resolve( preferred( ...args ) )
+			.then( res => res === undefined ? boundFallback() : res, boundFallback );
```

Comment:
I don't think we should treat an error the same as no result, since then there'd be no way for a function in the middle to indicate there was a problem – I think errors should fall through